libraries/program-error/tests/mod.rs

```diff
@@ -0,0 +1,109 @@
+pub mod bench;
+pub mod decode;
+pub mod into;
+pub mod print;
+pub mod spl;
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use solana_program::{
+        decode_error::DecodeError,
+        program_error::{PrintProgramError, ProgramError},
+    };
+
+    // `#[derive(IntoProgramError)]`
+    #[test]
+    fn test_derive_into_program_error() {
+        // `Into<ProgramError>`
+        assert_eq!(
+            Into::<ProgramError>::into(bench::ExampleError::MintHasNoMintAuthority),
+            Into::<ProgramError>::into(into::ExampleError::MintHasNoMintAuthority),
+        );
+        assert_eq!(
+            Into::<ProgramError>::into(bench::ExampleError::IncorrectMintAuthority),
+            Into::<ProgramError>::into(into::ExampleError::IncorrectMintAuthority),
+        );
+    }
+
+    // `#[derive(DecodeError)]`
+    #[test]
+    fn test_derive_decode_error() {
+        // `Into<ProgramError>`
+        assert_eq!(
+            Into::<ProgramError>::into(bench::ExampleError::MintHasNoMintAuthority),
+            Into::<ProgramError>::into(decode::ExampleError::MintHasNoMintAuthority),
+        );
+        assert_eq!(
+            Into::<ProgramError>::into(bench::ExampleError::IncorrectMintAuthority),
+            Into::<ProgramError>::into(decode::ExampleError::IncorrectMintAuthority),
+        );
+        // `DecodeError<T>`
+        assert_eq!(
+            <bench::ExampleError as DecodeError<bench::ExampleError>>::type_of(),
+            <bench::ExampleError as DecodeError<decode::ExampleError>>::type_of(),
+        );
+    }
+
+    // `#[derive(PrintProgramError)]`
+    #[test]
+    fn test_derive_print_program_error() {
+        // `Into<ProgramError>`
+        assert_eq!(
+            Into::<ProgramError>::into(bench::ExampleError::MintHasNoMintAuthority),
+            Into::<ProgramError>::into(print::ExampleError::MintHasNoMintAuthority),
+        );
+        assert_eq!(
+            Into::<ProgramError>::into(bench::ExampleError::IncorrectMintAuthority),
+            Into::<ProgramError>::into(print::ExampleError::IncorrectMintAuthority),
+        );
+        // `DecodeError<T>`
+        assert_eq!(
+            <bench::ExampleError as DecodeError<bench::ExampleError>>::type_of(),
+            <bench::ExampleError as DecodeError<print::ExampleError>>::type_of(),
+        );
+        // `PrintProgramError`
+        // (!) Not sure how better to test this yet - thoughts?
```

Comment:
There *is* a way, but it's a little annoying.

Since these are calling `msg!` off-chain, that'll default to `println!` based on the default program stubs at https://github.com/solana-labs/solana/blob/0495051a673b0599385891bb2b5caaf650eab879/sdk/program/src/program_stubs.rs#L27

If you want to test this properly, you can implement your own type which implements `SyscallStubs` for `sol_log` which writes the string to some mutable static variable, and then read out that string. That's how we used to test programs back before program-test!

The token program does something similar to this for return data, so you should be able to copy that: https://github.com/solana-labs/solana-program-library/blob/f1620caeaaf739f21ef788e7a19662b8e62c2009/token/program/src/processor.rs#L1055

Reply:
Working on this - last piece